src/Bicep.Cli/Commands/WhatIfCommand.cs

```diff
@@ -0,0 +1,140 @@
+        BicepCompiler compiler,
+        IOContext io,
+        ILogger logger,
+        IDeploymentManagerFactory deploymentManagerFactory,
+        IConfigurationManager configurationManager)
+    {
+        this.diagnosticLogger = diagnosticLogger;
+        this.compiler = compiler;
+        this.deploymentManagerFactory = deploymentManagerFactory;
+        this.configurationManager = configurationManager;
+        this.io = io;
+        this.logger = logger;
+    }
+
+    public async Task<int> RunAsync(WhatIfArguments args, CancellationToken cancellationToken)
+    {
+        var deploymentFile = ArgumentHelper.GetFileUri(args.InputFile);
+        ArgumentHelper.ValidateBicepDeployFile(deploymentFile);
+        var compilation = await compiler.CreateCompilation(
+            deploymentFile,
+            skipRestore: args.NoRestore);
+
+        var summary = diagnosticLogger.LogDiagnostics(DiagnosticOptions.Default, compilation);
+
+        // TODO: Use bicepdeploy compilation emission here
+        var result = compilation.Emitter.Template();
+
+        if (result.Template is not { } template)
+        {
+            return 1;
+        }
+
+        if (!compilation.GetEntrypointSemanticModel().Features.DeploymentFileEnabled)
+        {
+            logger.LogError("Experimental feature 'deployment file' is not enabled.");
+            return 1;
+        }
+
+        // TODO: Use deploy file emission to build deployment definition
+        var deploymentDefinition = new ArmDeploymentDefinition(
+            null,
+            "92722693-40f1-44fe-8c39-9cf6b6353750",
+            "levi-bicep-deploy",
+            args.Name ?? "main",
+            new ArmDeploymentWhatIfProperties(ArmDeploymentMode.Incremental)
+            {
+                Template = new BinaryData(JsonDocument.Parse(template).RootElement),
+                Parameters = new BinaryData("{}"),
+                WhatIfResultFormat = args.ResultFormat
+            });
+
+        var rootConfiguration = configurationManager.GetConfiguration(deploymentFile);
+        var deploymentManager = deploymentManagerFactory.CreateDeploymentManager(rootConfiguration);
+
+        try
+        {
+            var whatIfResult = await deploymentManager.WhatIfAsync(deploymentDefinition, cancellationToken);
+
+            await WriteWhatIfSummary(whatIfResult);
+        }
+        catch (WhatIfException ex)
+        {
+            await io.Error.WriteLineAsync($"Unable to run what-if: {ex.Message}");
+            return 1;
+        }
+        
+        return 0;
+    }
+
+    private async Task WriteWhatIfSummary(WhatIfOperationResult whatIfResult)
+    {
+        if (whatIfResult.Error is { } error)
+        {
+            await io.Error.WriteLineAsync($"What-if failed: {error.Code} - {error.Message}");
+        }
+
+        await io.Output.WriteLineAsync($"WhatIf status: {whatIfResult.Status}");
+        
+        // TODO: there's probably a better way to print these out
```

Comment:
We should port https://github.com/Azure/azure-powershell/blob/main/src/Resources/ResourceManager/Formatters/WhatIfOperationResultFormatter.cs, but that's out of scope of the hackathon project.

Reply:
Added TODO